### ISSUE TITLE
PHP 7 - Wrong Vector/Map/Set equality

### ISSUE THREAD

**Issue by tyx** [php-ds/extension#67]
I think creating a different issue from #65 is a better idea.

# Current problem
```
var_dump(new \Ds\Set([1, 3]) == new \Ds\Set([1, 2, 3])); // true
```

it makes me very sad as I definitively can't rely on Ds in my tests. And I use `Ds` everywhere as it does a really nice job (thanks for that anyway)

# Solution

Is there any reason why this behavior ? And is it a bug or a known behavior ? Looks too big that anyone complain about it ^^

Whatever I'm definitively open to contribute event if my C skill are very old because it cannot stay like that.

I already had a look on the (clean) code. 
I guess we need to implements the [compare_objects](https://wiki.php.net/internals/engine/objects#compare_objects) handlers on the internal htable ? Looks pretty obvious, I know. May be we can even use the `toArray` and rely on `array::compare_objects` ?
Is there any stuff that prevent us to do that ?

Anyway, let me know what I can do to help on this very annoying issue.

Thanks

**Comment by rtheunissen** [comment]
@tyx we would need to implement `compare_objects` on the zend objects, ie. `php_ds_set` etc. The internal htable would do the actual work though, eg.:

- `php_ds_set_t` `compare_objects`
- `php_ds_set_equals_object(php_ds_set_t *set, zend_object *obj)`
- Check if the object is an instance of Set
- Acces the internal set
- `ds_set_equals(ds_set_t *a, ds_set_t* b)`
- `ds_htable_equals(ds_htable_t *a, ds_htable_t *b)`

I'm guessing we'd iterate through both the table's keys and values and check if they're equal? (Or in the case of a Set, only the keys as all the values would be undefined).

`toArray` would be a performance hit, might also run into non-scalar keys.

Sorry about this annoying issue - keen to get it implemented and tested though.

There's also the obvious win if the object being compared to is the same instance. ⚡️

**Comment by tyx** [comment]
Looks nice ! Thanks you for your fast answer.

> toArray would be a performance hit, might also run into non-scalar keys.
Indeed ! 

> Sorry about this annoying issue - keen to get it implemented and tested though.
Compared to the benefits we get to use your extension, it's not a big deal in fact ;)

> There's also the obvious win if the object being compared to is the same instance.
Yes ! This one works actually

**Comment by rtheunissen** [comment]
@nikic would this actually be implicit if we implement the `get_properties` handler?

**Comment by nikic** [comment]
@rtheunissen Yes-ish. I'd assume you wouldn't want to have the same comparison semantics as a property comparison would give you. In particular, those would be weak and I doubt you'd want to have `Set(["foo"]) == Set([true])`, right?

**Comment by rtheunissen** [comment]
The only issue here is that we can't implement this in the polyfill. :no_good_man:

**Comment by rtheunissen** [comment]
Only option would be to implement `Hashable` and use `equals` for equality checks. Would never need `>` and `<` anyway.

**Comment by shouze** [comment]
@rtheunissen ok so the main issue is equality so it's not a big issue about `<` and `>`. And yes why not with `Hashable`, we're ok that Set, Vector, ... don't implement it ATM? This is what you've planned to solve the issue.

**Comment by rtheunissen** [comment]
@shouze happy to do it with equals. Added to 2.0, see #65

**Comment by rtheunissen** [comment]
This is actually more difficult than I expected. `equals` as part of the `Hashable` interface sort of scopes that equality check to a context in which the object is used as a key in a hash table. I'm not sure if that also implies general equality, ie. a case where you'd check `$a === $b`. 

For example, if we implement `equals` on `Collection`, and we're checking each item of the collection for strict equality, do we call `equals` on values that implement `Hashable`?

Or should we create an `Equality` interface and have `Hashable` just extend that?

**Comment by shouze** [comment]
First of all, to me identity comparison (`===`) is something else and not part of the scope of this issue.

> For example, if we implement equals on Collection, and we're checking each item of the collection for strict equality, do we call equals on values that implement Hashable?

Yes why not.

> Or should we create an Equality interface and have `Hashable` just extend that?

And yes why not, I'm not sure that it's needed but could be a nice to have.



In fact I wonder about the way you want to proceed:
1. do you want to lookup each entries of eg a `Collection` when we want to check that a collection is equal to another?
2. Or do you plan to maintain an always up2date internal hash, updated each time we `mutate` the collection?

**Comment by rtheunissen** [comment]
> Do you want to lookup each entries of eg a Collection when we want to check that a collection is equal to another?

Yes I think so. Something similar to what lodash does with _.equals. We can fail fast with things like type and size comparisons. An interesting case would be a Map, which may not have the same ordering, but could be considered "equal". `{a: 1, b: 2} == {b: 2, a: 1}` for example. We would have to define this for each collection.

> Or do you plan to maintain an always up2date internal hash, updated each time we mutate the collection?

No I don't think this is practical, better to just iterate through to compare.

**Comment by rtheunissen** [comment]
The issue remains that we can't implement a `==` check in the polyfill. Whenver you would use `$set == $another`, you would have to use `$set.equals($another)`.

**Comment by shouze** [comment]
@rtheunissen yes ok for `$set->equals($another)` if it can help to maintain the polyfill at the same level of features alongside.

**Comment by rtheunissen** [comment]
I'd like to keep them consistent, otherwise it invalidates the polyfill to some extent. Maybe if and when the extension becomes a default, we can drop the polyfill.

**Comment by shouze** [comment]
Yes, can be a scheduled deprecation but you're right it's probably not the time to do that as it would break adoption.

**Comment by rtheunissen** [comment]
This change is actually quite complex. Whenever values are compared (`filter`, `sort`, `find`, etc), we now have to check if the value implements `Hashable`, and use its `Hashable::equals` if it does. Of course it's easy to write a helper like `valuesAreEqual` that checks this for us, but the effects are clearly outside of the `Hashable` domain. I'm therefore convinced that we should move the `equals` method of `Hashable` to its own interface, maybe `Equality`, with only a single method `equals`. 

It's a shame that PHP still doesn't have a `Comparable` interface or operator overloading. I'm tempted to wait for [the Comparable RFC](https://wiki.php.net/rfc/comparable) to pass and be released, which would make the `Equality` interface redundant.

You can use something like `$a->diff($b)->isEmpty()` but that wouldn't currently call `equals` on any objects in the set that implement `Hashable`. If you know you only have scalar values in your set, you can wrap it and implement your own `equals` for now.

**Comment by marcospassos** [comment]
We're are considering to replace our custom collection structures by DS, but we use collection comparisons heavily. 

Java defines specific strategies for computing hash codes for each type of collection. For maps, for instance, the hash code is order insensitive `(hash += hash(key) ^ has(value))`, so that the equality relation is valid across all map implementations.

Our collection library has an `Equality` interface, from which `Hashable` extends from. It makes sense for us since the concept of hash only makes sense tied to an equality relation. Defining a concept of equality relation between collection also requires defining how the hash codes should be computed consistently as per collection type (it should be part of the interface contract). In Java, for instance, two maps are equal if both have the same EntrySet, while two sets are equal if both have the same elements. The rule for determining if two objects are equals is as follows:

- Two identical objects are always equal
- Two Hashable objects $a and $bare equal if $a->equals($b) and $b->equals($a)
- Two objects are equal if $a === $b

Despite the issues related to the implementation of operation comparison overloading, it would be really helpful to have a `equals` method that allows us testing collection for equality. It's is not a BC break and does not require complex changes as implementing operation overloading does.

About the Comparable RFC, it's an old discussion and I don't believe it will be added to the core anytime soon.

**Comment by rtheunissen** [comment]
> It's is not a BC break and does not require complex changes as implementing operation overloading does.

That's true, but changing the behaviour of `diff`, `intersect`, `remove`, etc could be considered significant enough to add as part of 2.0 (end 2017).

I agree that the collections should implement Hashable, and I think following the direction you've laid out makes sense. No need for the overloaded op just yet, that can come later.

**Comment by marcospassos** [comment]
@rtheunissen thanks for considering it! Indeed, changing the behavior of these methods does represent a BC break, but introducing the equals method does not. Any chance of having it before the 2.0 release? Looking the issues, it is probably the most requested feature.

**Comment by rtheunissen** [comment]
So we're implementing an equals method, but not Hashable yet?

**Comment by marcospassos** [comment]
Why not Hashable? We can just leave `diff`, `intersect`, `remove` untouched until 2.0. For now, it should be possible to implement `equals` and `hash` without introducing a BC break, right?

**Comment by rtheunissen** [comment]
But then Map and Set will try to use them as Hashable's, and not `spl_object_hash`, which could break things. We can add an `equals` method that defines equality, but implementing Hashable is a 2.0 addition.

**Comment by marcospassos** [comment]
Right, I missed it. Yes, `equals` will help a lot for now. Just to be sure, how do you plan to implement this algorithm? Performance is always a problem in collection comparisons. What I've in mind is something like the following algorithm (but obviously in C):

```php
public function equals($other): bool
{
    if (!($other instanceof Map)){
        return false;
    }

    if ($this->count() !== $other->count()) {
        return false;
    }

    foreach ($this->map as $key => $value) {
        if (!$this->map->hasKey($key)) {
            return false;
        }

        /**
         * isEqual
         * - Two identical objects are always equal
         * - Two Hashable objects $a and $b, are equal if $a->equals($b) and $b->equals($a)
         * - Everything else is unequal
         */

        if (!$this->isEqual($value, $this->map->get($key))) {
            return false;
        }
    }

    return true;
}
```

**Comment by rtheunissen** [comment]
So `{a: 1, b:2} === {b:2, a: 1}` ? I'm considering making the comparison order-sensitive, because it's an ordered collection, unlike a Java Map (excl ordered map).

**Comment by marcospassos** [comment]
From Java docs:

> Compares the specified object with this map for equality. Returns true if the given object is also a map and the two maps represent the same mappings. More formally, two maps m1 and m2 represent the same mappings if m1.entrySet().equals(m2.entrySet()). This ensures that the equals method works properly across different implementations of the Map interface.

So, in Java ordered and unordered collections and considered equal. 

However, I like your idea. We could further improve the collection by introducing a `containsAll(Map $map)`, so that `$left->count() === $right->count() && $left->containsAll($right)` is the same of a order insensitive equality relation. In that way we provide a way for performing both checks.

**Comment by rtheunissen** [comment]
What if we added a bool arg to `xor`, `diff` and `intersect` to also match on the value. That way, to determine if a Map is equal you can check if `count($map1->xor($map2, true)) == 0` ? I think that works?

**Comment by rtheunissen** [comment]
Can always have a bool arg in `equals`, which won't break the Hashable contract. `bool $ordered = false`? I really don't like boolean arguments though because their function is not always clear from the caller's end.

**Comment by rtheunissen** [comment]
I think we should keep it simple here. Check all pairs and make the same ordering a requirement for equality. If you have a potentially different order, sort first then check check equality.

**Comment by marcospassos** [comment]
I think it can quickly become unmaintainable. So, the behavior of these flags set to false is the same as not implementing the `Hashable` interface. In other words, if an object implements the `Hashable` interface it is explicitly defining the standard way for comparing two instances of that type. I think that we should enforce adherence to the protocol. Furthermore, if one wants this behavior and there is no big perfomance gain between the native array and the map, there is no reason for using the map at all.

**Comment by rtheunissen** [comment]
Java's argument for not considering ordering a part of the criteria for equality is based on the fact that their collections aren't ordered in the first place (bar a few). What's interesting to note is that Java doesn't consider ordering even when using collections like `LinkedMap`. See http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/util/LinkedHashMap.java

**Comment by marcospassos** [comment]
Yes, as I previously noticed:

> So, in Java ordered and unordered collections and considered equal.

Another point to consider is how to compare arrays. This is my suggestion:

```
class Foo implements Hashable {
    private $name;
    public function equals($b) {
         return $b instanceof Hashable && $this->name === $b->name;
    }
}

$left = new Map();
$left->set(0, [new Foo('a')]);

$right = new Map();
$right->set(0, [new Foo('a')]);

assertTrue($left->equals($right));

```

**Comment by rtheunissen** [comment]
What does the PHP array do? https://3v4l.org/bbTb0

Order is important when considering equality of an ordered collection. Java is wrong here IMO.

**Comment by rtheunissen** [comment]
> Another point to consider is how to compare arrays

Oh no.. we may have a problem here. If it's an array we'll need to traverse it? Recursively? See that's where this gets very messy where we don't have a native compareTo.

**Comment by marcospassos** [comment]
But it brings inconsistency. Is there any way of including the `Hashable::equals` as part of the `===` comparison operator? If so, it could be achieved transparently.

**Comment by rtheunissen** [comment]
I know that you can at least make it work with `==`, but I'm not sure about `===`. We'd need to diverge from the polyfill to support that though. `===` should be reserved for "same instance / reference" checks, so really all we need is the `==`.

**Comment by marcospassos** [comment]
`==` makes much more sense. The good part about that is that we don't need to care about comparing array, it would be done indirectly.

**Comment by rtheunissen** [comment]
Exactly, no need to implement `equals` at all. The issue lies in the polyfill, which we're considering to abandon for 2.0

**Comment by rtheunissen** [comment]
If we implement == in 1.x, we have an inconsistency with the polyfill.

**Comment by marcospassos** [comment]
I don't know how hard is it, but if it was possible to implement for `DateTime` it should not be hard to implement against an interface (or at least I hope so :)

**Comment by marcospassos** [comment]
Constraining the roadmap according to a polyfill is a big mistake IMHO. The polyfill makes no sense for production use. For any other purpose, I believe that an IDE plugin our just stubs are the way to go.

**Comment by marcospassos** [comment]
We have another problem. If we use the equality operator, then type juggling will break our plans. Furthermore, the operator `==` is order insensitive in regards to arrays. So, is there any way to reuse the strict comparison mechanism to compare arrays considering `Hashable`?

**Comment by rtheunissen** [comment]
> Furthermore, the operator == is order insensitive in regards to arrays

Not true! https://3v4l.org/bbTb0

**Comment by rtheunissen** [comment]
> type juggling will break our plans

How so?

**Comment by marcospassos** [comment]
From the docs:

Example | Name | Result
-- | -- | --
$a + $b | Union | Union of $a and $b.
$a == $b | Equality | TRUE if $a and $b have the same key/value pairs.
$a === $b | Identity | TRUE if $a and $b have the same key/value pairs in the same order and of the same types.
$a != $b | Inequality | TRUE if $a is not equal to $b.
$a <> $b | Inequality | TRUE if $a is not equal to $b.
$a !== $b | Non-identity | TRUE if $a is not identical to $b.

http://php.net/manual/en/language.operators.array.php

> > type juggling will break our plans
> How so?

`[1, true, false] == ["1", 1, 0]`

**Comment by rtheunissen** [comment]
Ahh... well don't use arrays then. ;)

**Comment by marcospassos** [comment]
It's not a choice. There is no way to constrain it.

**Comment by rtheunissen** [comment]
I stand corrected: https://3v4l.org/jdIBv

**Comment by marcospassos** [comment]
Yeah, maybe the behavior changed in PHP 7 but the doc is outdated. Anyway, the major issue here is the type juggling.

> So, is there any way to reuse the strict comparison mechanism to compare arrays considering Hashable? It would make the operator itself untouched, but it could be used by the `equals`method.

?

**Comment by rtheunissen** [comment]
I think we should try to stay as close to arrays as possible. 😞 

- Ordering doesn't matter
- `==` comparison on everything?

**Comment by marcospassos** [comment]
I personally don't use the operator `==`. I try to avoid it as much as possible. 

> I think we should try to stay as close to arrays as possible. 😞

Stay close to array implies that order matters. About the equality relation, just close the eyes for arrays is a big inconsistency IMHO. Resolving it put this library to the next level 🚀

**Comment by rtheunissen** [comment]
> Stay close to array implies that order matters

I thought we just agreed that arrays don't consider order important?

`[0 => 'a', 1 => 'b'] == [1 => 'b', 0 => 'a']`

**Comment by marcospassos** [comment]
In the context of arrays, the order is relevant. For loose comparison, it is not. However, an `equals` method can never rely on type juggling, it makes no sense. The only way I see it working is performing a strict comparison, exactly as `===` does, with an exception only for `Hashable` objects. This logic must be applied to array elements as well. IMHO, it's the cleanest and safest approach.

**Comment by rtheunissen** [comment]
Damn it, PHP. Not making this easy for us. 😠  But there's a solution here, just have to find it.

**Comment by marcospassos** [comment]
How about storing arrays in the Map as Maps, transparently to the developer? When an array is put into a Map, it became immutable as you cannot pass it as a reference. Storing it as a Map allows us labeling it with a hash code which eliminates most part of the inequalities. For the other cases, we can use the comparison logic recursively, calling the `equals` method on the wrapped array.

**Comment by rtheunissen** [comment]
Not a crazy idea, but I don't like the idea of putting an array in and getting a collection out.

**Comment by rtheunissen** [comment]
Let's just say that arrays use === comparison, end of story. If you need proper equality comparison, use a collection. *or* we can recursively descend into arrays and use the same equality checks. But that feels inconsistent and unexpected.

**Comment by marcospassos** [comment]
> Not a crazy idea, but I don't like the idea of putting an array in and getting a collection out.

This is why I said it should be handled transparently to the developer, so an array should be returned.

> Let's just say that arrays use === comparison, end of story. If you need proper equality comparison, use a collection. or we can recursively descend into arrays and use the same equality checks. But that feels inconsistent and unexpected.
Ignoring hashable inside arrays feels inconsistent as well :)

**Comment by rtheunissen** [comment]
```
$ php -a
Interactive shell

php > var_dump(['a' => 1, 'b' => 2] === ['b' => 2, 'a' => 1]);
bool(false)
php > var_dump(['a' => 1, 'b' => 2] === ['a' => 1, 'b' => 2]);
bool(true)
php > var_dump([false, true] === [0, 1]);
bool(true)
```

Here's the policy:

- Objects are compared with `equals` or `===` if they don't implement Equality.
- Everything else is compared with `===`. 
- **Objects within arrays that implement Equality will honour the interface.**

The third point here is the important one. We don't need to do anything clever or hacky to make this work, all we need to do is define the internal comparison behaviour. 

The only cost of doing this is that we can't support it in the polyfill, which supports the argument that the polyfill is not of practical value and should be abandoned.

**Comment by marcospassos** [comment]
@rtheunissen 100% agree with you. Any chance of getting this anytime soon?

**Comment by rtheunissen** [comment]
@marcospassos I don't want to release it under 1.x, but I could implement this and only this on a new branch? And merge that into 2.0 later on. But I have a lot on my plate at the moment so can't guarantee anything soon.

Do you need `equals` more than you need them to be hashable also? If that's the case, could you create your own collection that uses Map as the base and implement `equals` yourself in the meantime?

**Comment by marcospassos** [comment]
Do you use Skype, Slack or any other messaging app? (I will delete this message)

**Comment by rtheunissen** [comment]
Gitter?

**Comment by rtheunissen** [comment]
@marcospassos https://gitter.im/php-ds/extension

**Comment by Fleshgrinder** [comment]
Interesting to see the typical PHP issues that internals claims are no issues being discussed here.

Regarding your discussion of what is equal and what is identity. I usually go for two different interfaces:

```php
interface Identifiable {
    function idents($other): bool; // ===
    function hash(): string; // spl_object_hash
}

interface Equalable {
    function equals($other): bool; // ==
}
```

The identifiable interface is for [equivalence relations](https://en.wikipedia.org/wiki/Equivalence_relation) and the equalable interface is for [partial equivalence relations](https://en.wikipedia.org/wiki/Partial_equivalence_relation). This also matches what proper programming languages do (e.g. Rust or Ceylon).

The thing regarding order is actually very simple from a math point of view. Two collections (set or map) are equal if they contain the same elements, regardless of order. This is why every proper language provides different types for e.g. set and sorted set as well as map and sorted map.

Another thing that is very important to understand is that equality is not covered by comparability. The same ordering of something does not directly translate to equality. I explained this once in internals but the best and easiest example is `INF` (or `NAN`).

Infinity is not smaller nor greater than infinity, however, infinity does not equal infinity.

```
INF.equals(INF) // false
INF.compareTo(INF) // 0
```

There are more examples with date and time but I think this is sufficient as an explanation. Hence yet another of my typical interfaces:

```php
interface Comparable {
    /** @param static $other */
    function compareTo($other): Ordering;
}
```

Also, just for your consideration, providing methods to call is faster than overloading the operators (a lot if you have many comparisons). I benchmarked this once with a C implementation where I allowed overloading for all operators with marker interface as the ones I showed here.

**Comment by rtheunissen** [comment]
Coming back to the plan of action for 2.0 here:
- Override `==` for all collections to determine equality
- Have `===` compare object reference
- Implement `Hashable` in all collections, where `equals` simply compares using `==`.

Consequences:
- No need to separate `equals` from `Hashable`
- Abandon the polyfill 🔪

**Comment by rtheunissen** [comment]
Hi everyone. :wave: 

My plan for comparison behavior in 2.0:
- Abandon the polyfill 🔪
- Hash-based structures only support object-keys that implement Hashable.
- ` interface Hashable extends Equatable`
- Overload `==` for **all** classes that implement `Equatable`. 
- Use the equivalent of `===` for everything else (scalars, arrays, etc).
- Two objects can only be equal if they are also instances of the same class.
- The only Hashable collection will be `Tuple`, which is immutable.

**Comment by shouze** [comment]
@rtheunissen this is the right decision to me. I've abandoned this extension because of this kind of issues and honestly if all of that is because of the polyfill kill it!

**Comment by hopeseekr** [comment]
It's such a shame to abandon the polyfill.

**Comment by eduardbareev** [comment]
> I'm considering making the comparison order-sensitive, because it's an ordered collection, unlike a 

Hi. Please consider providing an option for order-insensitive maps comparison. The current order-sensitive behavior `===` for dict-arrays is impractical.

It is more problem of default map in php being ordered though. In many cases insertion order carries no meaning if you look at how given dict was instantiated, let alone those being a result of merge of several other dicts, or populated based on data-source with no stable order.

Python might be an interesting case with regard to dict comparison. While being factually ordered, officially dict was unordered until 3.7. Therefore equality test was order-insensitive, as it would be nonsense to take physical order into account for officially unordered collection.

At 3.7 dict became officially insertion-order. However, dict comparison ignores the order (officially so). 
```
>>> {'baz': 1, 'foo': 2} == {'foo': 2, 'baz': 1}
True
```
Now the collection is ordered but the comparison is not. 

IMO this turns to be very useful behavior for some applications.

Another aspect is interoperability. Assuming maps in the wild to be ordered has no basis as the prevalent serialization format is JSON (which is essentially spec for literals syntax). Preserving order would require a counterpart system to use an order-preserving json decoder and encoder and prefer, say, LinkedHashMap to HashMap. The same is with storage. PostgreSQL `jsonb` type doesn't preserve order of json. As of related type which does they say: "quite specialized needs, such as legacy assumptions about ordering of object keys".

An example of PHP's `===` for dict-arrays being useless is an automated test ensuring that retrieved map is "equal" to the one previously persisted. As of `==`, it is not more useful.

Following are 3-tuples of dict, list and set for which equality test in python yields true.
```
>>> ({'foo': 2, 'baz': 1}, [1,2], {1,2}) == ({'baz': 1, 'foo': 2}, [1,2], {2,1})
True
```
Nested one:
```
{'k': {'k': {'k': {'k': {'k': 'v', 'k2': 'v2'}}}}} == {'k': {'k': {'k': {'k': {'k2': 'v2', 'k': 'v'}}}}}
True
```
Keys and values comparison follows the `==`, which is equality test, i.e. `__eq__()` call. Keys always implement `__hash__`, because they are not accepted as keys otherwise, thus they implement `__eq__`. Values compared equality-wise, too. In case value doesn't implement `__eq__`, parent `object.__eq__` performs an identity test as a last resort. As of comparing primitive to object, `{'k': Foo()} == {'k': 123}`, should `Foo.__eq__` decide so. But that's because python have no value types that is, `int` is reference type too. So it is more or less equals/hashCode minus the fact that there is no value types.

For me not having "intuitive" operators for collections doesn't seem to be a problem. After all, in Java `==` is an identity test and `String` is reference type and people somehow got used to it (while in C# it was fixed by making `==` equality for strings). Having sane and useful `.equals()` (and acceptance of equals/hashCode protocol) seems more important IMHO.

- https://docs.python.org/3/reference/datamodel.html#object.__hash__
- https://docs.python.org/3.7/library/stdtypes.html#dict.values
